Serialize and deserialize resizable ArrayBuffer

Constellation opened the pull request:
#### a9510aedf5179f92aa036b67aea7f5f9e8b0a24f
<pre>
Serialize and deserialize resizable ArrayBuffer
<a href="https://bugs.webkit.org/show_bug.cgi?id=248209">https://bugs.webkit.org/show_bug.cgi?id=248209</a>
rdar://102601423

Reviewed by Ross Kirsling.

This patch adds serializing and deserializing of resizable ArrayBuffer and TypedArrays.
We add ResizableArrayBufferTag and add a feature serializing resizable ArrayBuffer.
But for growable SharedArrayBuffer, nothing is necessary since information is carried via
SharedArrayBufferContents already. For TypedArrays, we use UINT64_MAX byteLength marker
as a auto-length case. This works since byteLength cannot be UINT64_MAX since it exceeds
MAX_ARRAY_BUFFER_SIZE. The other things in TypedArrays are not changed much since these
TypedArrays should be resizable / growable ones when the subsequent backing serialized
ArrayBuffer is resizable.

We also add wrappedAs methods since normal tryCreate has more additional checks for construction.
But these checks can fail if the serialized TypedArrays are having ArrayBuffers which is resized
to be smaller after the construction. But this is OK since it just makes TypedArrays OOB. wrappedAs
methods do not have this check.

* Source/JavaScriptCore/runtime/ArrayBufferView.h:
(JSC::ArrayBufferView::byteOffsetRaw const):
(JSC::ArrayBufferView::byteOffset const):
(JSC::ArrayBufferView::byteLengthRaw const):
(JSC::ArrayBufferView::byteLength const):
* Source/WebCore/bindings/js/SerializedScriptValue.cpp:
(WebCore::CloneSerializer::dumpArrayBufferView):
(WebCore::CloneSerializer::dumpIfTerminal):
(WebCore::CloneDeserializer::readResizableNonSharedArrayBuffer):
(WebCore::CloneDeserializer::readArrayBufferViewImpl):
(WebCore::CloneDeserializer::readTerminal):

Canonical link: <a href="https://commits.webkit.org/256998@main">https://commits.webkit.org/256998@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a195e21dd66fc96d8494e198a864d2ac66d57590

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/97510 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/6765 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/30675 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/107015 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/167278 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/101467 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/7100 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/35521 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/89905 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/103684 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/103165 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/5320 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/84141 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/32326 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/87193 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/88982 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/75251 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/88417 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/768 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/20446 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/84083 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/756 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/21922 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/28493 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4821 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/5569 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/44396 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/86910 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/2006 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/41296 "Passed tests") | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/19515 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

rkirsling commented on Source/WebCore/bindings/js/SerializedScriptValue.cpp:
```diff
@@ -3956,6 +4010,23 @@ class CloneDeserializer : CloneBase {
+            Structure* structure = m_globalObject->arrayBufferStructure(arrayBuffer->sharingMode());
+            // A crazy RuntimeFlags mismatch could mean that we are not equipped to handle shared
+            // array buffers while the sender is. In that case, we would see a null structure here.
+            if (!structure) {
```
UNLIKELY? Since you said it'd be a "crazy mismatch".

Constellation replied:
Changed.

rkirsling reviewed (approved):
LGTM, but does this functionality have any upstream tests available?

Constellation committed:
Serialize and deserialize resizable ArrayBuffer
https://bugs.webkit.org/show_bug.cgi?id=248209
rdar://102601423

Reviewed by Ross Kirsling.

This patch adds serializing and deserializing of resizable ArrayBuffer and TypedArrays.
We add ResizableArrayBufferTag and add a feature serializing resizable ArrayBuffer.
But for growable SharedArrayBuffer, nothing is necessary since information is carried via
SharedArrayBufferContents already. For TypedArrays, we use UINT64_MAX byteLength marker
as a auto-length case. This works since byteLength cannot be UINT64_MAX since it exceeds
MAX_ARRAY_BUFFER_SIZE. The other things in TypedArrays are not changed much since these
TypedArrays should be resizable / growable ones when the subsequent backing serialized
ArrayBuffer is resizable.

We also add wrappedAs methods since normal tryCreate has more additional checks for construction.
But these checks can fail if the serialized TypedArrays are having ArrayBuffers which is resized
to be smaller after the construction. But this is OK since it just makes TypedArrays OOB. wrappedAs
methods do not have this check.

* Source/JavaScriptCore/runtime/ArrayBufferView.h:
(JSC::ArrayBufferView::byteOffsetRaw const):
(JSC::ArrayBufferView::byteOffset const):
(JSC::ArrayBufferView::byteLengthRaw const):
(JSC::ArrayBufferView::byteLength const):
* Source/WebCore/bindings/js/SerializedScriptValue.cpp:
(WebCore::CloneSerializer::dumpArrayBufferView):
(WebCore::CloneSerializer::dumpIfTerminal):
(WebCore::CloneDeserializer::readResizableNonSharedArrayBuffer):
(WebCore::CloneDeserializer::readArrayBufferViewImpl):
(WebCore::CloneDeserializer::readTerminal):

Canonical link: https://commits.webkit.org/256998@main
